maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

```diff
@@ -123,31 +130,166 @@ public fun GoogleMap(
+        return getTag(R.id.maps_compose_map_view_tag_click_listeners) as? MapClickListeners ?: MapClickListeners()
+            .also { clickListeners ->
+                setTag(R.id.maps_compose_map_view_tag_click_listeners, clickListeners)
+            }
+    }
+
+    /** Apply the [content] compositions to the map. */
+    suspend fun setComposition(mapView: MapView) {
+        mapView.log("setComposition")
+        val clickListeners = mapView.getOrCreateMapListenersTag()
+        val (currentComposition, mapReused) = mapView.getOrCreateComposition(clickListeners)
+
+        composition = currentComposition
+
+        val mapCompositionContent: @Composable () -> Unit = {
+            MapUpdater(
+                mergeDescendants = mergeDescendants,
+                contentDescription = currentContentDescription,
+                cameraPositionState = currentCameraPositionState,
+                contentPadding = currentContentPadding,
+                locationSource = currentLocationSource,
+                mapProperties = currentMapProperties,
+                mapUiSettings = currentUiSettings,
+            )
+
+            MapClickListenerUpdater()
+
+            CompositionLocalProvider(
+                LocalCameraPositionState provides currentCameraPositionState,
+            ) {
+                currentContent?.invoke()
+            }
+        }
+
+        currentComposition.apply {
+            if(mapReused) {
+                setContentWithReuse(mapCompositionContent)
+            } else {
+                setContent(mapCompositionContent)
+            }
+        }
+
+        // Set this after composition is started.
+        mapClickListeners = clickListeners
+    }
+
+    AndroidView(
+        modifier = modifier,
+        factory = {
+            Log.d(TAG, "Factory")
+            debugMapReused = false
+            MapView(context, googleMapOptionsFactory()).also { mapView ->
+                mapLifecycleController = MapViewLifecycleController(
+                    isMapReused = false,
+                    mapView = mapView
+                )
+                componentCallbacks = mapView.componentCallbacks()
+            }
+        },
+        onReset = { mapView ->
+            mapView.log("onReset")
+            // Deactivate composition to save resources
+            context.unregisterComponentCallbacks(componentCallbacks)
+            mapLifecycleController!!.onLifecycleDetached()
+            composition?.deactivate()
+            // Call onStop/onPause or something? Set map type to None to save resources? Because the MapView is detached.
+        },
+        onRelease = { mapView ->
+            mapView.log("onRelease")
+            context.unregisterComponentCallbacks(componentCallbacks)
+            // Dispose composition
+            composition?.dispose()
+            // Invoke onDestroy + remove lifecycle callbacks for the MapView
+            mapLifecycleController!!.onDestroy()
```

Comment:
Lifecycle issue is fixed.
Seems good to not touch MapType.

I store the context for which the componentCallbacks are registered to and check the context on each reuse to make sure that the context is the same. Or else I re-register the componentCallbacks. Not sure that it's necessary. What do you think?

Reply:
Making special provisions for `Context` changes looks unnecessary, because: `Context` is only passed into the `AndroidView` `factory` lambda, not `update` or `onReset` lambdas. If `Context` changed, everything would break. Pretty sure Views are tied to a specific context, one can't just swap out the Context.

Just need to make sure it's unregistered in `onRelease()` and/or `onDestroy()`.

Reply:
True. I've removed that code.

Should be enough to unregister in `onRelease` because it's supposed to be used to clean up resources used by the `View`.